src/PostgresOutbox/Subscriptions/Subscription.cs

```diff
@@ -0,0 +1,160 @@
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Microsoft.Extensions.Logging;
+using Npgsql;
+using Npgsql.Replication;
+using Npgsql.Replication.PgOutput;
+using Npgsql.Replication.PgOutput.Messages;
+using PostgresOutbox.Database;
+using PostgresOutbox.Subscriptions.Management;
+using PostgresOutbox.Subscriptions.ReplicationMessageHandlers;
+using PostgresOutbox.Subscriptions.SnapshotReader;
+
+namespace PostgresOutbox.Subscriptions;
+
+using static PublicationManagement;
+using static ReplicationSlotManagement;
+using static ReplicationSlotManagement.CreateReplicationSlotResult;
+
+public interface ISubscription
+{
+    IAsyncEnumerable<IEnvelope> Subscribe(Func<SubscriptionOptionsBuilder, SubscriptionOptionsBuilder> builder,
+        ILoggerFactory? loggerFactory,
+        CancellationToken ct);
+}
+
+public enum CreateStyle
+{
+    WhenNotExists,
+    AlwaysRecreate,
+    Never
+}
+
+public sealed class Subscription: ISubscription, IAsyncDisposable
+{
+    private static LogicalReplicationConnection? _connection;
+    private static readonly SubscriptionOptionsBuilder Builder = new();
+    private ISubscriptionOptions? _options;
+    public async IAsyncEnumerable<IEnvelope> Subscribe(
+        Func<SubscriptionOptionsBuilder, SubscriptionOptionsBuilder> builder,
+        ILoggerFactory? loggerFactory = null,
+        [EnumeratorCancellation] CancellationToken ct = default
+    )
+    {
+        _options = builder(Builder).Build();
+        var (connectionString, publicationSetupOptions, slotSetupOptions, errorProcessor, replicationDataMapper, registry) = _options;
+        var dataSourceBuilder = new NpgsqlDataSourceBuilder(connectionString);
```

Comment:
**CONCERN:** Maybe we could also pass NpgsqlDataSource to the Subscribe method? That'd make easier global setup.

Reply:
Sure. Consider this PR a good enough starting point

Reply:
Added https://github.com/oskardudycz/PostgresOutboxPatternWithCDC.NET/issues/16